aws_lambda_builders/workflows/nodejs_npm_esbuild/workflow.py

```diff
@@ -53,150 +50,75 @@ def __init__(self, source_dir, artifacts_dir, scratch_dir, manifest_path, runtim
-
-        :type install_action: BaseAction
-        :param install_action: Installation action for npm
-
-        :type esbuild_with_deps: BaseAction
-        :param esbuild_with_deps: Standard esbuild action bundling source with deps
+        if self.download_dependencies:
+            self.actions += [
+                NodejsNpmWorkflow.get_install_action(
+                    source_dir=source_dir,
+                    artifacts_dir=self.scratch_dir,
+                    subprocess_npm=self.subprocess_npm,
+                    osutils=self.osutils,
+                    build_options=self.options,
+                ),
+                bundle_action,
+            ]
 
-        :type esbuild_no_deps: List[BaseAction]
-        :param esbuild_no_deps: esbuild action not including dependencies in the bundled artifacts
+        if not self.dependencies_dir:
+            # we need a dependencies dir for the logic below
+            return
 
-        :rtype: list
-        :return: List of build actions to execute
-        """
         if self.download_dependencies:
-            actions += [install_action, CleanUpAction(self.dependencies_dir)]
-            if self.combine_dependencies:
-                # Auto dependency layer disabled, first build
-                actions += [
-                    esbuild_with_deps,
-                    MoveDependenciesAction(source_dir, scratch_dir, self.dependencies_dir),
-                ]
-            else:
-                # Auto dependency layer enabled, first build
-                actions += esbuild_no_deps + [MoveDependenciesAction(source_dir, scratch_dir, self.dependencies_dir)]
+            # if we downloaded dependencies, we have to update dependencies_dir
+            self.actions += [
+                CleanUpAction(self.dependencies_dir),
+                MoveDependenciesAction(self.source_dir, self.scratch_dir, self.dependencies_dir),
+            ]
         else:
-            if self.dependencies_dir:
-                actions.append(LinkSourceAction(self.dependencies_dir, scratch_dir))
-
-                if self.combine_dependencies:
-                    # Auto dependency layer disabled, subsequent builds
-                    actions += [esbuild_with_deps]
-                else:
-                    # Auto dependency layer enabled, subsequent builds
-                    actions += esbuild_no_deps
-            else:
-                # Invalid workflow, can't have no dependency dir and no installation
-                raise EsbuildExecutionError(
-                    message="Lambda Builders encountered an invalid workflow. A workflow can't "
-                    "include a dependencies directory without installing dependencies."
-                )
-
-        return actions
+            # if we're reusing dependencies, then we need to symlink them before bundling
+            self.actions += [
+                LinkSourceAction(self.dependencies_dir, self.scratch_dir),
+                bundle_action,
```

Comment:
Seems that the bundle action is already included in line 103, so this would run the bundler twice.

Reply:
This is one part that can lead to confusion and wasn't able to figure out how to make it better. It would never bundle twice because this bundle on this line happens when `self.download_dependencies` is `False`, while the one in line 103 happens when `self.download_dependencies` is `True`

Reply:
in other words, if we're reusing dependencies, then we didn't download them and we should first symlink and then bundle. If we downloaded dependencies, then we should bundle after downloading, (and we won't symlink or bundle again)

Reply:
I'm seeing it if it's true for both cases:
```
if self.download_dependencies:
            self.actions += [
                NodejsNpmWorkflow.get_install_action(
                    source_dir=source_dir,
                    artifacts_dir=self.scratch_dir,
                    subprocess_npm=self.subprocess_npm,
                    osutils=self.osutils,
                    build_options=self.options,
                ),
                bundle_action,
            ]

        if not self.dependencies_dir:
            # we need a dependencies dir for the logic below
            return

        if self.download_dependencies:
            # if we downloaded dependencies, we have to update dependencies_dir
            self.actions += [
                CleanUpAction(self.dependencies_dir),
                MoveDependenciesAction(self.source_dir, self.scratch_dir, self.dependencies_dir),
            ]
        else:
            # if we're reusing dependencies, then we need to symlink them before bundling
            self.actions += [
                LinkSourceAction(self.dependencies_dir, self.scratch_dir),
                bundle_action,
            ]
```

Reply:
I think you're indentation might be off? 😅 . The first if in the code block you're posting should be in the same level as the other two if's below

Reply:
or otherwise not sure what you mean

Reply:
I just made an additional change related to this, hopefully it's clearer now.